### PR TITLE
fix(mutate): fix types of mutate/trigger; make mutate/trigger always return the result of fetcher

### PR DESCRIPTION
**Pull request by Ponyets**
Fix for [issue 2690](https://github.com/vercel/swr/issues/2690) [issue 2647](https://github.com/vercel/swr/issues/2647) and [issue 2661](https://github.com/vercel/swr/issues/2661)

1. Make mutate/trigger always return the result of fetcher
2. Make KeyedMutator/ScopedMutator accept generic type parameters to make the data type flexible and work with populateCache

### CODE REVIEW

**Comment by Ponyets** [comment]
Hello. 
It seems this PR need an approval to run the test workflow, since I had to modify a test file.
Could a maintainer kindly give it a try? Thanks.

**Comment by Ponyets** [comment]
I ran `pnpm run-all-checks` locally this time. It passed.
Fixed `FIXME` left by @shuding in `test/type/mutate.ts`.

File: _internal/src/types.ts [hunk 1]
```diff
@@ -310,14 +310,17 @@ export type MutatorCallback<Data = any> = (
   currentData?: Data
 ) => Promise<undefined | Data> | undefined | Data
 
-export type MutatorOptions<Data = any> = {
+export type MutatorOptions<SWRData = any, Data = SWRData> = {
```
Comment (shuding) [inline _internal/src/types.ts:313]:
```suggestion
export type MutatorOptions<Data = any, MutationData = Data> = {
```

Maybe this naming is better? I think we’d like to have a JSDoc here explaining what these generics mean actually. Still kinda confusing.

**Review by shuding — COMMENTED** [review]
Thanks for the PR, it looks great in general!

File: _internal/src/types.ts (hunk 1 above)
Comment (Ponyets) [reply _internal/src/types.ts]: Updated as your suggestion, JSDoc added as well.

**Pull request merged**